### ISSUE TITLE
Immutable root state

### ISSUE THREAD

**Issue by fabriziomoscon** [redux-offline/redux-offline#37]
Hi @sorodrigo,
thanks a lot for opening and maintaining this repository.
A couple of months ago I opened this PR: https://github.com/jevakallio/redux-offline/pull/99
to allow the use of `immutable root state` with `redux-offline`.
Since then we have been successfully using the fork with our immutable root state.
I wanted to check with you whether the same PR could be merged in this fork if I create it and rebase it in top of the current improvements.

Thanks

**Comment by sebasgarcep** [comment]
@fabriziomoscon Could you link your fork to review it?

**Comment by sorodrigo** [comment]
Sounds interesting! I skimmed through the other PR and through redux-offline-immutable-config. The only thing I'm not thrilled about, is that even if one's not dealing with an immutable store, there's a slight overhead introduced to the core because of the lenses. 
Now, don't get me wrong, immutable support is a nice thing to have! I'd like to hear more from @wacii given that he's been updating the enhacer and other internals recently.

**Comment by sorodrigo** [comment]
On another note, the redux-offline org seems like a good place where your config package could live in.

**Comment by wacii** [comment]
It looks good. About as compact a change as I think possible.

Injecting `autoRehydrate()` is something we should be doing already. It doesn't make a whole lot of sense to inject `persist()` but not `autoRehydrate()` as we do now.

As for the lens, you are either going to need such an abstraction or to build two versions of the project. I prefer adding an abstraction.

**Comment by wacii** [comment]
That said I don't understand the benefit of making the root object immutable.

**Comment by fabriziomoscon** [comment]
Thanks for all comments.
The original idea was to make 100% non breaking changes. For that we set defaults everywhere possible.
To allow the use of `immutable.js` we had to inject:
- persistCallback
- persistAutoRehydrate
- offlineStateLens
but defaults will allow user to ignore them.

When using immutable setting the entire redux state as immutable is a best practice:
http://redux.js.org/docs/recipes/UsingImmutableJS.html#make-your-entire-redux-state-tree-an-immutablejs-object

`redux-offline-immutable-config` provides default configuration for users using `immutable.js`, we created a separate repo so that only immutable users need to install the extra bits.

I will go through the rebase and see if there are any incompatibilities
